README.md
# text_table

Renders style-able flex ascii tables.

# Features

+ Customizable table borders
+ Multiline rows
+ Column size control
    + Min and max Column widths
    + Flex column widths
    + Percentage based column widths
+ Cell text alignment
+ Cell padding

# TODO

+ [ ] Merged cells
+ [ ] Text truncation and ellipsis

# Usage

## Simple example

```dart
import 'package:text_table/text_table.dart';

main() {
  final tab = TableRenderer().render([
    ['Messi', 'Barcelona FC', 80],
    ['Christiano Ronaldo', 'Real Madrid', 30],
    ['Luiz Suarez', 'Barcelona FC', 50]
  ], columns: [
    'Player',
    'Team',
    'Goals'
  ]);
  print(tab);
}
```

```
┌──────────────────┬────────────┬─────┐
│      Player      │    Team    │Goals│
┝━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━┿━━━━━┥
│Messi             │Barcelona FC│80   │
├──────────────────┼────────────┼─────┤
│Christiano Ronaldo│Real Madrid │30   │
├──────────────────┼────────────┼─────┤
│Luiz Suarez       │Barcelona FC│50   │
└──────────────────┴────────────┴─────┘
```

## Text alignment

```dart
final tab = TableRenderer().render([
  ['Messi', 'Barcelona FC', 80],
  ['Christiano Ronaldo', 'Real Madrid', 30],
  ['Luiz Suarez', 'Barcelona FC', 50]
], columns: [
  ColSpec(name: 'Player', width: Fixed(25), align: Align.left),
  ColSpec(name: 'Team', width: Fixed(25), align: Align.center),
  ColSpec(name: 'Goals', width: Fixed(10), align: Align.right)
]);
print(tab);
```

```
┌─────────────────────────┬─────────────────────────┬──────────┐
│         Player          │          Team           │  Goals   │
┝━━━━━━━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━┥
│Messi                    │      Barcelona FC       │        80│
├─────────────────────────┼─────────────────────────┼──────────┤
│Christiano Ronaldo       │       Real Madrid       │        30│
├─────────────────────────┼─────────────────────────┼──────────┤
│Luiz Suarez              │      Barcelona FC       │        50│
└─────────────────────────┴─────────────────────────┴──────────┘
```

## Padding

```dart
import 'package:text_table/text_table.dart';

main() {
  final tab = TableRenderer(padding: Padding.same(1)).render([
    ['Messi', 'Barcelona FC', 80],
    ['Christiano Ronaldo', 'Real Madrid', 30],
    ['Luiz Suarez', 'Barcelona FC', 50]
  ], columns: [
    'Player',
    'Team',
    ColSpec(name: 'Goals', padding: Padding.same(2)),
  ]);
  print(tab);
}
```

```
┌────────────────────┬──────────────┬─────────┐
│       Player       │     Team     │  Goals  │
┝━━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━━━┿━━━━━━━━━┥
│ Messi              │ Barcelona FC │  80     │
├────────────────────┼──────────────┼─────────┤
│ Christiano Ronaldo │ Real Madrid  │  30     │
├────────────────────┼──────────────┼─────────┤
│ Luiz Suarez        │ Barcelona FC │  50     │
└────────────────────┴──────────────┴─────────┘
```

## Flex columns

```dart
import 'package:text_table/text_table.dart';

main() {
  final tab = TableRenderer(padding: Padding.same(2)).render([
    ['Messi', 'Barcelona FC', 80],
    ['Christiano Ronaldo', 'Real Madrid', 30],
    ['Luiz Suarez', 'Barcelona FC', 50]
  ], columns: [
    ColSpec(name: 'Player', width: Flex(2)),
    ColSpec(name: 'Team', width: Flex(1), minWidth: Fixed(15)),
    ColSpec(name: 'Goals', width: Flex(1), align: Align.right)
  ], width: 50);
  print(tab);
}
```

```
┌─────────────────┬───────────────────┬──────────┐
│     Player      │       Team        │  Goals   │
┝━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━┥
│  Messi          │  Barcelona FC     │      80  │
├─────────────────┼───────────────────┼──────────┤
│  Christiano Ro  │  Real Madrid      │      30  │
│  naldo          │                   │          │
├─────────────────┼───────────────────┼──────────┤
│  Luiz Suarez    │  Barcelona FC     │      50  │
└─────────────────┴───────────────────┴──────────┘
```

## Customize border styles

Use `border` field of `TableRenderer` to customize the border style of the table.

### Double line

```dart
final tab = TableRenderer(border: Border.doubleLines)
    .render(rows, columns: columns);
print(tab);
```

```
╔══════════════════╤════════════╤═════╗
║      Player      │    Team    │Goals║
╠══════════════════╪════════════╪═════╣
║Messi             │Barcelona FC│80   ║
╟──────────────────┼────────────┼─────╢
║Christiano Ronaldo│Real Madrid │30   ║
╟──────────────────┼────────────┼─────╢
║Luiz Suarez       │Barcelona FC│50   ║
╚══════════════════╧════════════╧═════╝
```

### Compact

```dart
final tab =
  TableRenderer(border: Border.compact).render(rows, columns: columns);
print(tab);
```

```
      Player           Team     Goals
-------------------------------------
Messi              Barcelona FC 80   
Christiano Ronaldo Real Madrid  30   
Luiz Suarez        Barcelona FC 50    
```

### Rounded corners

```dart
final tab =
  TableRenderer(border: Border.round).render(rows, columns: columns);
print(tab);
```

```
╭──────────────────┬────────────┬─────╮
│      Player      │    Team    │Goals│
┝━━━━━━━━━━━━━━━━━━┿━━━━━━━━━━━━┿━━━━━┥
│Messi             │Barcelona FC│80   │
├──────────────────┼────────────┼─────┤
│Christiano Ronaldo│Real Madrid │30   │
├──────────────────┼────────────┼─────┤
│Luiz Suarez       │Barcelona FC│50   │
╰──────────────────┴────────────┴─────╯
```

CHANGELOG.md
# Changelog

## 4.0.3

+ Render last line bug fix

## 4.0.2

+ Updated some package versions

## 4.0.1

+ Fixed padding bug

## 4.0.0

**Complete Rewrite!**
+ `TableRenderer.render` renders the table
+ Introduced `ColSpec` class to control column properties like width, minimum
width, maximum width, alignment, padding, etc.

## 3.0.0

+ Null safety

## 2.6.0

+ Ported to Dart 2.0.0

## 2.5.4

+ Bug fix when there is only row 

## 2.5.3

+ Headless table

## 2.5.2

+ Fixes for Dart2

## 2.5.1

+ Padding
+ Flex tables
+ Column width customization
+ Column width estimation

## 1.0.0

+ Single line tables


example/table/align.dart
import 'package:text_table/text_table.dart';

main() {
  // Left align
  {
    final tab = TableRenderer().render([
      ['Messi', 'Barcelona FC', 80],
      ['Christiano Ronaldo', 'Real Madrid', 30],
      ['Luiz Suarez', 'Barcelona FC', 50]
    ], columns: [
      ColSpec(name: 'Player', width: Fixed(25)),
      ColSpec(name: 'Team', width: Fixed(25)),
      ColSpec(name: 'Goals', width: Fixed(10))
    ]);
    print(tab);
  }

  // Right align
  {
    final tab = TableRenderer().render([
      ['Messi', 'Barcelona FC', 80],
      ['Christiano Ronaldo', 'Real Madrid', 30],
      ['Luiz Suarez', 'Barcelona FC', 50]
    ], columns: [
      ColSpec(name: 'Player', width: Fixed(25), align: Align.right),
      ColSpec(name: 'Team', width: Fixed(25), align: Align.right),
      ColSpec(name: 'Goals', width: Fixed(10), align: Align.right)
    ]);
    print(tab);
  }

  // Center align
  {
    final tab = TableRenderer().render([
      ['Messi', 'Barcelona FC', 80],
      ['Christiano Ronaldo', 'Real Madrid', 30],
      ['Luiz Suarez', 'Barcelona FC', 50]
    ], columns: [
      ColSpec(name: 'Player', width: Fixed(25), align: Align.center),
      ColSpec(name: 'Team', width: Fixed(25), align: Align.center),
      ColSpec(name: 'Goals', width: Fixed(10), align: Align.center)
    ]);
    print(tab);
  }

  {
    final tab = TableRenderer().render([
      ['Messi', 'Barcelona FC', 80],
      ['Christiano Ronaldo', 'Real Madrid', 30],
      ['Luiz Suarez', 'Barcelona FC', 50]
    ], columns: [
      ColSpec(name: 'Player', width: Fixed(25), align: Align.left),
      ColSpec(name: 'Team', width: Fixed(25), align: Align.center),
      ColSpec(name: 'Goals', width: Fixed(10), align: Align.right)
    ]);
    print(tab);
  }
}
